src/psbt.c

```diff
@@ -0,0 +1,2387 @@
+#define TX_OUTPUT_ALLOC(typ) \
+    *output = wally_malloc(sizeof(typ)); \
+    if (!*output) return WALLY_ENOMEM; \
+    wally_clear((void *)*output, sizeof(typ)); \
+    result = (typ *) *output;
+
+static bool clone_bytes(unsigned char **dst, const unsigned char *src, size_t len)
+{
+    if (!len)
+        return true;
+    *dst = wally_malloc(len);
+    if (*dst)
+        memcpy(*dst, src, len);
+    return *dst != NULL;
+}
+
+static int clone_tx(struct wally_tx* tx, struct wally_tx **output)
+{
+    unsigned char *bytes;
+    size_t calc, written;
+    int ret = WALLY_OK;
+
+    TX_CHECK_OUTPUT;
+
+    if ((ret = wally_tx_get_length(tx, WALLY_TX_FLAG_USE_WITNESS, &calc)) != WALLY_OK) {
+        goto done;
+    }
+    if ((bytes = wally_malloc(calc)) == NULL) {
+        ret = WALLY_ENOMEM;
+        goto done;
+    }
+    if ((ret = wally_tx_to_bytes(tx, WALLY_TX_FLAG_USE_WITNESS, bytes, calc, &written)) != WALLY_OK) {
+        goto done;
+    }
+    if ((ret = wally_tx_from_bytes(bytes, written, WALLY_TX_FLAG_USE_WITNESS, output)) != WALLY_OK) {
+        goto done;
+    }
+done:
+    if (bytes) {
+        wally_free(bytes);
+    }
+    return ret;
+}
+
+static struct wally_tx_witness_stack *clone_witness(
+    const struct wally_tx_witness_stack *stack)
+{
+    struct wally_tx_witness_stack *result;
+    size_t i;
+    int ret;
+
+    ret = wally_tx_witness_stack_init_alloc(stack->items_allocation_len, &result);
+
+    if (ret == WALLY_OK) {
+        for (i = 0; i < stack->num_items && ret == WALLY_OK; ++i) {
+            if (stack->items[i].witness) {
+                ret = wally_tx_witness_stack_set(result, i,
+                                                 stack->items[i].witness,
+                                                 stack->items[i].witness_len);
+                if (ret != WALLY_OK)
+                    wally_tx_witness_stack_free(result);
+            }
+        }
+    }
+    return ret == WALLY_OK ? result : NULL;
+}
+
+
+static void clear_and_free(void *p, size_t len)
+{
+    if (p) {
+        wally_clear(p, len);
+        wally_free(p);
+    }
+}
+
+static bool pubkey_is_compressed(unsigned char pubkey[65]) {
+    return pubkey[0] == 0x02 || pubkey[0] == 0x03;
+}
+
+
+int keypath_map_init_alloc(size_t alloc_len, struct keypath_map **output)
+{
+    struct keypath_map *result;
+
+    TX_CHECK_OUTPUT;
+    TX_OUTPUT_ALLOC(struct keypath_map);
+
+    if (alloc_len) {
+        result->items = wally_malloc(alloc_len * sizeof(*result->items));
+        if (!result->items) {
+            wally_free(result);
+            *output = NULL;
+            return WALLY_ENOMEM;
+        }
+        wally_clear(result->items, alloc_len * sizeof(*result->items));
+    }
+    result->items_allocation_len = alloc_len;
+    result->num_items = 0;
+    return WALLY_OK;
+}
+
+int keypath_map_free(struct keypath_map *keypaths)
+{
+    size_t i;
+
+    if (keypaths && keypaths->items) {
```

Comment:
Might it be better to make clearing/freeing keypaths only dependent on keypaths. It looks like if you call keypath_map_init_alloc with alloc_len=0 it will allocate a keypath_map with no items which then won't be freed by keypath_map_free.

Reply:
I removed the check for items here and in the other map free functions.